libxrdp/xrdp_sec.c

```diff
@@ -2298,159 +2298,190 @@ xrdp_sec_process_mcs_data_channels(struct xrdp_sec *self, struct stream *s)
+            description->session_height > 0x7FFE || description->session_height < 0xC8)
     {
-        LOG(LOG_LEVEL_ERROR,
+        LOG(LOG_LEVEL_INFO,
             "Client supplied virtual desktop width or height is invalid. "
             "Allowed width range: min %d, max %d. Width received: %d. "
             "Allowed height range: min %d, max %d. Height received: %d",
-            0xC8, 0x7FFE, client_info->width,
-            0xC8, 0x7FFE, client_info->height);
-        return 1; /* error */
+            0xC8, 0x7FFE, description->session_width,
+            0xC8, 0x7FFE, description->session_width);
+        goto exit_error;
     }
 
     /* keep a copy of non negative monitor info values for xrdp_wm usage */
-    for (index = 0; index < monitorCount; index++)
+    for (int monitor_index = 0; monitor_index < NumMonitor; ++monitor_index)
     {
-        client_info->minfo_wm[index].left =  client_info->minfo[index].left - x1;
-        client_info->minfo_wm[index].top =  client_info->minfo[index].top - y1;
-        client_info->minfo_wm[index].right =  client_info->minfo[index].right - x1;
-        client_info->minfo_wm[index].bottom =  client_info->minfo[index].bottom - y1;
-        client_info->minfo_wm[index].is_primary =  client_info->minfo[index].is_primary;
+        monitor_layout = description->minfo_wm + monitor_index;
+
+        g_memcpy(monitor_layout, description->minfo + monitor_index, sizeof(struct monitor_info));
+
+        monitor_layout->left = monitor_layout->left - rect.left;
+        monitor_layout->top = monitor_layout->top - rect.top;
+        monitor_layout->right = monitor_layout->right - rect.left;
+        monitor_layout->bottom = monitor_layout->bottom - rect.top;
     }
+    return description;
+exit_error:
+    g_free(description);
+    return NULL;
+}
+
+/*****************************************************************************/
+/* Process a [MS-RDPBCGR] TS_UD_CS_MONITOR message.
+   reads the client monitors data */
+static int
+xrdp_sec_process_mcs_data_monitors(struct xrdp_sec *self, struct stream *s)
+{
+    int flags;
+    struct xrdp_client_info *client_info = &(self->rdp_layer->client_info);
+
+    /* this is an option set in xrdp.ini */
+    if (client_info->multimon != 1) /* are multi-monitors allowed ? */
+    {
+         LOG(LOG_LEVEL_INFO, "Multi-monitor is disabled by server config");
+         return 0;
+    }
+    if (!s_check_rem_and_log(s, 8, "Parsing [MS-RDPBCGR] TS_UD_CS_MONITOR"))
+    {
+         return 1;
+    }
+    in_uint32_le(s, flags); /* flags */
+
+    //verify flags - must be 0x0
+    if (flags != 0)
+    {
+        LOG(LOG_LEVEL_ERROR,
+            "[MS-RDPBCGR] Protocol error: TS_UD_CS_MONITOR flags MUST be zero, "
+            "received: 0x%8.8x", flags);
+        return 1;
+    }
+
+    struct display_size_description *description = process_monitor_stream(s, 0);
+
+    client_info->monitorCount = description->monitorCount;
```

Comment:
Needs test for NULL return here if an invalid PDU is received.

Reply:
Good catch. Fixed.